RAP for CSS 4.4 opibuilder

A build of the RAP product against the CSS 4.4 p2 repository doesn't start with an error
> Caused by: java.lang.ClassNotFoundException: org.eclipse.swt.SWTError cannot be found by org.eclipse.ui.workbench_3.107.1.v20160120-2131

*swt probably shouldn't be there, so I compared the installed plugins and I found a number of potentially important differences between 4.3 and 4.4.

> org.eclipse.swt
> org.eclipse.swt.gtk.linux.x86
> org.eclipse.text
> org.eclipse.ui
> org.eclipse.ui.ide
> org.eclipse.ui.workbench

are in the 4.4 war, but not in 4.3.

While trying to find out what is pulled in first in the dependency chain, I came across this commit as a possible source: https://github.com/ControlSystemStudio/cs-studio/commit/79b95e1cb0bd95495c3dfdd6c7c78d50869db252. This brings org.eclipse.ui.ide to the opibuilder dependency list when before it was only used by the editor.

--- a/applications/opibuilder/opibuilder-plugins/org.csstudio.opibuilder.rap/src/org/csstudio/opibuilder/util/ResourceUtilSSHelperImpl.java
+++ b/applications/opibuilder/opibuilder-plugins/org.csstudio.opibuilder.rap/src/org/csstudio/opibuilder/util/ResourceUtilSSHelperImpl.java
@@ -190,4 +190,9 @@ public class ResourceUtilSSHelperImpl extends ResourceUtilSSHelper {
     public Image getScreenShotImage(GraphicalViewer viewer) {
         throw new RuntimeException(NOT_IMPLEMENTED);
     }
+
+    @Override
+    public IEditorInput editorInputFromPath(IPath path) {
+        throw new RuntimeException(NOT_IMPLEMENTED);
+    }
 }

--- a/applications/opibuilder/opibuilder-plugins/org.csstudio.opibuilder.rcp/src/org/csstudio/opibuilder/util/ResourceUtilSSHelperImpl.java
+++ b/applications/opibuilder/opibuilder-plugins/org.csstudio.opibuilder.rcp/src/org/csstudio/opibuilder/util/ResourceUtilSSHelperImpl.java
@@ -18,6 +18,8 @@ import java.util.logging.Logger;
 
 import org.csstudio.opibuilder.OPIBuilderPlugin;
 import org.csstudio.ui.util.NoResourceEditorInput;
+import org.eclipse.core.filesystem.EFS;
+import org.eclipse.core.filesystem.IFileStore;
 import org.eclipse.core.filesystem.URIUtil;
 import org.eclipse.core.resources.IFile;
 import org.eclipse.core.resources.IResource;
@@ -241,4 +243,21 @@ public class ResourceUtilSSHelperImpl extends ResourceUtilSSHelper {
         gc.dispose();
         return image;
     }
+
+    public IEditorInput editorInputFromPath(IPath path) {
+        IEditorInput editorInput = null;
+        IFile file = ResourcesPlugin.getWorkspace().getRoot().getFile(path);
+        // Files outside the workspace are handled differently
+        // by Eclipse.
+        if (!ResourceUtil.isExistingWorkspaceFile(path)
+                && ResourceUtil.isExistingLocalFile(path)) {
+            IFileStore fileStore = EFS.getLocalFileSystem()
+                    .getStore(file.getFullPath());
+            editorInput = new FileStoreEditorInput(fileStore);
+        } else {
+            editorInput = new FileEditorInput(file);
+        }
+        return editorInput;
+    }
+
 }

--- a/applications/opibuilder/opibuilder-plugins/org.csstudio.opibuilder/META-INF/MANIFEST.MF
+++ b/applications/opibuilder/opibuilder-plugins/org.csstudio.opibuilder/META-INF/MANIFEST.MF
@@ -35,10 +35,7 @@ Require-Bundle: org.eclipse.ui;resolution:=optional,
  org.python.jython;bundle-version="2.5.2",
  org.csstudio.swt.widgets;bundle-version="2.1.0",
  org.csstudio.swt.xygraph;bundle-version="2.0.4",
- org.eclipse.e4.ui.workbench,
- org.eclipse.core.filesystem,
- org.eclipse.core.resources,
- org.eclipse.ui.ide
+ org.eclipse.e4.ui.workbench
 Eclipse-RegisterBuddy: org.python.jython
 Bundle-ActivationPolicy: lazy
 Export-Package: org.csstudio.opibuilder,

--- a/applications/opibuilder/opibuilder-plugins/org.csstudio.opibuilder/src/org/csstudio/opibuilder/util/ResourceUtil.java
+++ b/applications/opibuilder/opibuilder-plugins/org.csstudio.opibuilder/src/org/csstudio/opibuilder/util/ResourceUtil.java
@@ -35,10 +35,6 @@ import org.csstudio.opibuilder.OPIBuilderPlugin;
 import org.csstudio.opibuilder.model.AbstractWidgetModel;
 import org.csstudio.opibuilder.persistence.URLPath;
 import org.csstudio.opibuilder.preferences.PreferencesHelper;
-import org.eclipse.core.filesystem.EFS;
-import org.eclipse.core.filesystem.IFileStore;
-import org.eclipse.core.resources.IFile;
-import org.eclipse.core.resources.ResourcesPlugin;
 import org.eclipse.core.runtime.IPath;
 import org.eclipse.core.runtime.IProgressMonitor;
 import org.eclipse.core.runtime.IStatus;
@@ -53,8 +49,6 @@ import org.eclipse.swt.graphics.Image;
 import org.eclipse.swt.graphics.ImageData;
 import org.eclipse.swt.graphics.ImageLoader;
 import org.eclipse.ui.IEditorInput;
-import org.eclipse.ui.ide.FileStoreEditorInput;
-import org.eclipse.ui.part.FileEditorInput;
 
 /**Utility functions for resources.
  * @author Xihui Chen, Abadie Lana, Kay Kasemir
@@ -429,19 +423,7 @@ public class ResourceUtil {
     }
 
     public static IEditorInput editorInputFromPath(IPath path) {
-        IEditorInput editorInput = null;
-        IFile file = ResourcesPlugin.getWorkspace().getRoot().getFile(path);
-        // Files outside the workspace are handled differently
-        // by Eclipse.
-        if (!ResourceUtil.isExistingWorkspaceFile(path)
-                && ResourceUtil.isExistingLocalFile(path)) {
-            IFileStore fileStore = EFS.getLocalFileSystem()
-                    .getStore(file.getFullPath());
-            editorInput = new FileStoreEditorInput(fileStore);
-        } else {
-            editorInput = new FileEditorInput(file);
-        }
-        return editorInput;
+        return IMPL.editorInputFromPath(path);
     }
 
     /**Get the first existing file on search path. Search path is a BOY preference.

--- a/applications/opibuilder/opibuilder-plugins/org.csstudio.opibuilder/src/org/csstudio/opibuilder/util/ResourceUtilSSHelper.java
+++ b/applications/opibuilder/opibuilder-plugins/org.csstudio.opibuilder/src/org/csstudio/opibuilder/util/ResourceUtilSSHelper.java
@@ -84,9 +84,8 @@ public abstract class ResourceUtilSSHelper {
      */
     public abstract IPath workspacePathToSysPath(IPath path);
 
-
     public abstract Image getScreenShotImage(GraphicalViewer viewer);
 
-
+    public abstract IEditorInput editorInputFromPath(IPath path);
 
 }

--- a/applications/scan/scan-plugins/org.csstudio.scan.diirt.datasource/src/org/csstudio/scan/diirt/datasource/ScanChannelHandler.java
+++ b/applications/scan/scan-plugins/org.csstudio.scan.diirt.datasource/src/org/csstudio/scan/diirt/datasource/ScanChannelHandler.java
@@ -42,13 +42,13 @@ import org.diirt.vtype.table.Column;
  * @author Eric Berryman
  */
 class ScanChannelHandler extends MultiplexedChannelHandler<ScanChannelHandler.ConnectionPayload, Object> {
-    
+
     private final ScanDataSource datasource;
     private final URI uri;
-    private final Map<String,String> queryMap = new ConcurrentHashMap<String,String>(); 
+    private final Map<String,String> queryMap = new ConcurrentHashMap<String,String>();
     private final Long id;
     private final REQUEST_TYPE requestType;
-    
+
     ScanChannelHandler(ScanDataSource datasource, URI uri, Long id, REQUEST_TYPE requestType) {
         super(uri.toASCIIString());
         this.datasource = datasource;
@@ -57,7 +57,7 @@ class ScanChannelHandler extends MultiplexedChannelHandler<ScanChannelHandler.Co
         this.requestType = requestType;
         this.queryMap.putAll(getQueryMap(uri.getQuery()));
     }
-    
+
     ScanChannelHandler(ScanDataSource datasource, URI uri, REQUEST_TYPE requestType) {
         super(uri.toASCIIString());
         this.datasource = datasource;
@@ -66,7 +66,7 @@ class ScanChannelHandler extends MultiplexedChannelHandler<ScanChannelHandler.Co
         this.requestType = requestType;
         this.queryMap.putAll(getQueryMap(uri.getQuery()));
     }
-    
+
     @Override
     public void connect() {
         poll();
@@ -81,7 +81,7 @@ class ScanChannelHandler extends MultiplexedChannelHandler<ScanChannelHandler.Co
     protected boolean isConnected(ScanChannelHandler.ConnectionPayload payload) {
         return payload != null && payload.connected;
     }
-    
+
     @Override
     protected boolean isWriteConnected(ScanChannelHandler.ConnectionPayload payload) {
         return requestType == REQUEST_TYPE.SCAN_INFO;
@@ -113,7 +113,7 @@ class ScanChannelHandler extends MultiplexedChannelHandler<ScanChannelHandler.Co
         }
         callback.channelWritten(null);
     }
-    
+
     static class ConnectionPayload {
         final boolean connected;
         final boolean pollQuerySuccessful;
@@ -125,22 +125,22 @@ class ScanChannelHandler extends MultiplexedChannelHandler<ScanChannelHandler.Co
             this.dataQuerySuccessful = dataQuerySuccessful;
         }
     }
-    
+
     private static final Object NO_POLL_DATA = new Object();
     private volatile Object pollResult;
-    
+
     void poll() {
         // Skip poll if channel is no usage on the channel
         if (getUsageCounter() > 0) {
             boolean connected = false;
             boolean pollQuerySuccessful = false;
             boolean dataQuerySuccessful = false;
-            
+
             // Retrieve scan connection
             try {
                 ScanClient scanClient = datasource.getConnection(uri.getHost());
                 connected = true;
-                
+
                 // Execute the poll and compare with the old value
                 Object newPollResult = executePollQuery(scanClient);
                 if (shouldReadData(pollResult, newPollResult)) {
@@ -181,7 +181,7 @@ class ScanChannelHandler extends MultiplexedChannelHandler<ScanChannelHandler.Co
             }
         }
     }
-    
+
     private static Map<String, String> getQueryMap(String query) {
         if(query == null){
             return new ConcurrentHashMap<String, String>();
@@ -195,11 +195,11 @@ class ScanChannelHandler extends MultiplexedChannelHandler<ScanChannelHandler.Co
         }
         return map;
     }
-    
+
     private boolean shouldReadData(Object oldPollResult, Object newPollResult) {
         return !Objects.equals(oldPollResult, newPollResult);
     }
-    
+
     private Object executePollQuery(ScanClient scanClient) throws Exception {
         if(requestType == REQUEST_TYPE.SERVER_INFO){
             return 1;
@@ -212,12 +212,12 @@ class ScanChannelHandler extends MultiplexedChannelHandler<ScanChannelHandler.Co
             }
         }
     }
-    
+
     // This is awful
     private VTable executeDataQuery(ScanClient scanClient) throws Exception{
         ScanData scanData = scanClient.getScanData(id);
         final Comparator<ScanSample> comp = (p1, p2) -> p1.getTimestamp().compareTo(p2.getTimestamp());;
-        
+
         List<Class<?>> types = new ArrayList<Class<?>>();
         List<String> names = new ArrayList<String>();
         List<Object> values = new ArrayList<Object>();
@@ -245,148 +245,148 @@ class ScanChannelHandler extends MultiplexedChannelHandler<ScanChannelHandler.Co
             names.add(device);
             values.add(new ArrayDouble(sampleData));
         }
-        
+
         return org.diirt.vtype.ValueFactory.newVTable(types, names, values);
     }
-    
+
     private VTable executeDevicesQuery(ScanClient scanClient) throws Exception{
         Collection<DeviceInfo> scanDevices = scanClient.getScanDevices(id);
         List<Column> columns = new ArrayList<Column>();
-        
+
         List<String> names = scanDevices.stream().map(deviceInfo -> deviceInfo.getName()).collect(Collectors.toList());
         columns.add(column("name", newVStringArray(names, alarmNone(),timeNow())));
-        
+
         List<String> aliases = scanDevices.stream().map(deviceInfo -> deviceInfo.getAlias()).collect(Collectors.toList());
         columns.add(column("alias", newVStringArray(aliases, alarmNone(),timeNow())));
-        
+
         List<String> status = scanDevices.stream().map(deviceInfo -> deviceInfo.getStatus()).collect(Collectors.toList());
         columns.add(column("status", newVStringArray(status, alarmNone(),timeNow())));
-        
+
         return newVTable(columns.toArray(new Column[columns.size()]));
     }
-    
+
     private VTable executeScanInfoQuery(ScanClient scanClient) throws Exception{
         ScanInfo scanInfo = scanClient.getScanInfo(id);
 
         List<Class<?>> types = new ArrayList<Class<?>>();
         List<String> names = new ArrayList<String>();
         List<Object> values = new ArrayList<Object>();
-        
+
         // id - possible overflow from long
         types.add(Integer.TYPE);
         names.add(Messages.Id);
         values.add(new ArrayInt((int)scanInfo.getId()));
-        
+
         // created
         types.add(Instant.class);
         names.add(Messages.Created);
         values.add(Arrays.asList(scanInfo.getCreated()));
-        
+
         // name
         types.add(String.class);
         names.add(Messages.Name);
         values.add(Arrays.asList(scanInfo.getName()));
-        
+
         // currentCommand
         types.add(String.class);
         names.add(Messages.CurrentCommand);
         values.add(Arrays.asList(scanInfo.getCurrentCommand()));
-        
+
         // finishTime
         types.add(Instant.class);
         names.add(Messages.FinishTime);
         values.add(Arrays.asList(scanInfo.getFinishTime()));
-        
+
         // percentage
         types.add(Integer.TYPE);
         names.add(Messages.Percentage);
         values.add(new ArrayInt(scanInfo.getPercentage()));
-        
+
         // state
         types.add(String.class);
         names.add(Messages.State);
         values.add(Arrays.asList(scanInfo.getState().toString()));
-        
+
         // error
         types.add(String.class);
         names.add(Messages.Error);
         values.add(Arrays.asList(scanInfo.getError().isPresent()?scanInfo.getError().get():""));
-        
+
         return org.diirt.vtype.ValueFactory.newVTable(types, names, values);
     }
-    
+
     private VTable executeServerInfoQuery(ScanClient scanClient) throws Exception{
         List<ScanInfo> scanInfos = scanClient.getScanInfos();
 
         int maxSize = Integer.parseInt(queryMap.getOrDefault("max", "1000"));
-        
+
         List<Class<?>> types = new ArrayList<Class<?>>();
         List<String> names = new ArrayList<String>();
         List<Object> values = new ArrayList<Object>();
-        
+
         // id - possible overflow from long
         types.add(Integer.TYPE);
         names.add(Messages.Id);
         int[] scanIds = scanInfos.stream().limit(maxSize).map(scanInfo -> (int)scanInfo.getId()).mapToInt(i -> i).toArray();
         values.add(new ArrayInt(scanIds));
-        
+
         // created
         types.add(Instant.class);
         names.add(Messages.Created);
         List<Instant> createds = scanInfos.stream().limit(maxSize).map(scanInfo -> scanInfo.getCreated()).collect(Collectors.toList());
         values.add(createds);
-        
+
         // name
         types.add(String.class);
         names.add(Messages.Name);
         List<String> scan_names = scanInfos.stream().limit(maxSize).map(scanInfo -> scanInfo.getName()).collect(Collectors.toList());
         values.add(scan_names);
-        
+
         // currentCommand
         types.add(String.class);
         names.add(Messages.CurrentCommand);
         List<String> currentCommands = scanInfos.stream().limit(maxSize).map(scanInfo -> scanInfo.getCurrentCommand()).collect(Collectors.toList());
         values.add(currentCommands);
-        
+
         // finishTime
         types.add(Instant.class);
         names.add(Messages.FinishTime);
         List<Instant> finishTimes = scanInfos.stream().limit(maxSize).map(scanInfo -> scanInfo.getFinishTime()).collect(Collectors.toList());
         values.add(finishTimes);
-        
+
         // percentage
         types.add(Integer.TYPE);
         names.add(Messages.Percentage);
         int[] percentages = scanInfos.stream().limit(maxSize).map(scanInfo -> scanInfo.getPercentage()).mapToInt(i -> i).toArray();;
         values.add(new ArrayInt(percentages));
-        
+
         // state
         types.add(String.class);
         names.add(Messages.State);
         List<String> states = scanInfos.stream().limit(maxSize).map(scanInfo -> scanInfo.getState().toString()).collect(Collectors.toList());
         values.add(states);
-        
+
         // error
         types.add(String.class);
         names.add(Messages.Error);
         List<String> errors = scanInfos.stream().limit(maxSize).map(scanInfo -> scanInfo.getError().isPresent()?scanInfo.getError().get():"").collect(Collectors.toList());
         values.add(errors);
-        
+
         return org.diirt.vtype.ValueFactory.newVTable(types, names, values);
     }
-    
+
     private void pause (ScanClient scanClient) throws Exception {
         scanClient.pauseScan(id);
     }
-    
+
     private void resume (ScanClient scanClient) throws Exception {
         scanClient.resumeScan(id);
     }
-    
+
     private void abort (ScanClient scanClient) throws Exception {
         scanClient.abortScan(id);
     }
-    
+
     private void remove (ScanClient scanClient) throws Exception {
         scanClient.removeScan(id);
     }

--- a/applications/scan/scan-plugins/org.csstudio.scan.diirt.datasource/src/org/csstudio/scan/diirt/datasource/ScanDataSource.java
+++ b/applications/scan/scan-plugins/org.csstudio.scan.diirt.datasource/src/org/csstudio/scan/diirt/datasource/ScanDataSource.java
@@ -29,13 +29,13 @@ import static org.diirt.util.concurrent.Executors.namedPool;
  *
  */
 public class ScanDataSource extends DataSource {
-    
+
     static {
         // Install type support for the types it generates.
         DataTypeSupport.install();
     }
-    
-    public enum REQUEST_TYPE { 
+
+    public enum REQUEST_TYPE {
         SCAN_INFO, SERVER_INFO, SCAN_DATA, SCAN_DEVICES
     }
 
@@ -47,7 +47,7 @@ public class ScanDataSource extends DataSource {
     private final Pattern scanInfoPath;
     private final Pattern dataPath;
     private final Pattern devicesPath;
-    
+
     static {
         // Install type support for the types it generates.
         DataTypeSupport.install();
@@ -60,20 +60,20 @@ public class ScanDataSource extends DataSource {
         scanInfoPath = Pattern.compile("\\/-?([0-9]+)\\/?");
         dataPath = Pattern.compile("\\/-?([0-9]+)\\/data");
         devicesPath = Pattern.compile("\\/-?([0-9]+)\\/devices");
-        
+
     }
 
     @Override
     public boolean isWriteable() {
         return true;
     }
-    
+
     @Override
     public void close() {
         exec.shutdown();
         super.close();
     }
-    
+
     private void poll() {
         for (ChannelHandler channel : getChannels().values()) {
             ((ScanChannelHandler)channel).poll();
@@ -85,7 +85,7 @@ public class ScanDataSource extends DataSource {
         try {
             // adding a protocol to simplify parsing
             URI uri = new URI("http://"+channelName);
-            
+
             if(configuration.connections.containsKey(uri.getHost())) {
                 if( uri.getPath().isEmpty() ){
                     log.fine("Creating Channel: "+channelName);
@@ -113,17 +113,17 @@ public class ScanDataSource extends DataSource {
                     log.fine("Creating Channel: "+channelName+" with id: "+String.valueOf(id));
                     return new ScanChannelHandler(this,uri,id, REQUEST_TYPE.SCAN_INFO);
                 }
-                            
+
                 throw new RuntimeException("Malformed URI scan channel named " + channelName);
             }
             throw new RuntimeException("Couldn't find scan channel named " + channelName);
         } catch (URISyntaxException e) {
             throw new RuntimeException("Malformed URI scan channel named " + channelName);
         }
-        
-        
+
+
     }
-    
+
     ScanClient getConnection(String connectionName) {
         ScanClient scanSource = scanSources.get(connectionName);
         if (scanSource == null) {

--- a/applications/scan/scan-plugins/org.csstudio.scan.diirt.datasource/src/org/csstudio/scan/diirt/datasource/ScanDataSourceConfiguration.java
+++ b/applications/scan/scan-plugins/org.csstudio.scan.diirt.datasource/src/org/csstudio/scan/diirt/datasource/ScanDataSourceConfiguration.java
@@ -44,21 +44,21 @@ public class ScanDataSourceConfiguration extends DataSourceConfiguration<ScanDat
             DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
             DocumentBuilder builder = factory.newDocumentBuilder();
             Document document = builder.parse(input);
-            
+
             XPathFactory xpathFactory = XPathFactory.newInstance();
             XPath xPath = xpathFactory.newXPath();
-            
+
             String ver = xPath.evaluate("/scan/@version", document);
             if (!ver.equals("1")) {
                 throw new IllegalArgumentException("Unsupported version " + ver);
             }
-            
+
             String pollIntervalSetting = xPath.evaluate("/scan/servers/@pollInterval", document);
             pollInterval = Integer.parseInt(pollIntervalSetting);
-            
+
             String timeOutSetting = xPath.evaluate("/scan/servers/@timeout", document);
             timeOut = Integer.parseInt(timeOutSetting);
-            
+
             Map<String, URL> newConnections = new HashMap<>();
             NodeList xmlChannels = (NodeList) xPath.evaluate("/scan/servers/server", document, XPathConstants.NODESET);
             for (int j = 0; j < xmlChannels.getLength(); j++) {
@@ -67,7 +67,7 @@ public class ScanDataSourceConfiguration extends DataSourceConfiguration<ScanDat
                 String url =  xPath.evaluate("@url", xmlChannel);
                 newConnections.put(channelName,new URL(url));
             }
-            
+
             connections = newConnections;
         } catch (ParserConfigurationException | SAXException | IOException | XPathExpressionException ex) {
             Logger.getLogger(ScanDataSourceConfiguration.class.getName()).log(Level.FINEST, "Couldn't load scan configuration", ex);

